image/store/store.go

```diff
@@ -0,0 +1,250 @@
+package store
+
+import (
+	"sync"
+
+	"github.com/alibaba/sealer/logger"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+
+	v1 "github.com/alibaba/sealer/types/api/v1"
+	"github.com/opencontainers/go-digest"
+)
+
+var imagestore *store
+var once sync.Once
+
+//StoreID is caculated from a series of serialized layers. The layers hash
+//value is "", but the layers value and type is not empty. It is used in
+//image store to help to reference to a stack of layers with a single identifier.
+type StoreID digest.Digest
+
+func (id StoreID) String() string {
+	return id.Digest().String()
+}
+
+// Digest converts ID into a digest
+func (id StoreID) Digest() digest.Digest {
+	return digest.Digest(id)
+}
+
+type ImageID digest.Digest
+
+func (id ImageID) String() string {
+	return id.Digest().String()
+}
+
+// Digest converts ID into a digest
+func (id ImageID) Digest() digest.Digest {
+	return digest.Digest(id)
+}
+
+// Store is an interface for manipulating images
+type Store interface {
+	Images() map[ImageID]*v1.Image
+	SetParent(id StoreID, parent StoreID) error
+	GetParent(id StoreID) (StoreID, error)
+	GetLayer(id StoreID) (v1.Layer, error)
+	Children(id StoreID) []StoreID
+}
+
+type imageMeta struct {
+	layer v1.Layer
+	// use store id
+	parent StoreID
+	// use store id
+	children map[StoreID]struct{}
+}
+
+type store struct {
+	sync.RWMutex
+	// use store id
+	images map[StoreID]*imageMeta
+	fs     StoreBackend
+	ls     LayerStore
+}
+
+func NewImageStore(fs StoreBackend, ls LayerStore) (Store, error) {
+	once.Do(func() {
+		imagestore = &store{
+			images: make(map[StoreID]*imageMeta),
+			fs:     fs,
+			ls:     ls,
+		}
+
+		if err := imagestore.restore(); err != nil {
+			return
+		}
+	})
+	return imagestore, nil
+}
+
+// restore reads all images saved in filesystem and calculate their storeid
+func (is *store) restore() error {
+	is.Lock()
+	defer is.Unlock()
+
+	//read all image layers
+	images := is.Images()
+	for _, image := range images {
+		layers := image.Spec.Layers
+		var tempLayers []v1.Layer
+		for i, layer := range layers {
+			tempLayers = append(tempLayers, v1.Layer{Value: layer.Value, Type: layer.Type})
+			image.Spec.Layers = tempLayers
+			// at present, we only calculate layers without hash
+			storeid, err := GetStoreIDFromImage(*image)
+			if err != nil {
+				logger.Error(err)
+				break
+			}
+
+			imagemeta, ok := is.images[storeid]
+			if !ok {
+				imagemeta = &imageMeta{
+					layer:    layer,
+					children: make(map[StoreID]struct{}),
+				}
+				is.images[storeid] = imagemeta
+			}
+
+			if i < len(layers)-1 {
+				image.Spec.Layers = append(tempLayers, v1.Layer{Value: layers[i+1].Value, Type: layers[i+1].Type})
```

Comment:
will it be nicer to set layers after for done, just use intermediate layer slice?